File: backend/app/services/agent_ui/bootstrap.py
from __future__ import annotations

from sqlalchemy.ext.asyncio import AsyncSession

from app.services.agent_ui.contracts import (
    AgentUiBootstrap,
    StarterPromptView,
    default_ui_capabilities,
)
from app.services.agent_ui.execution_targets import execution_target_catalog


async def build_agent_ui_bootstrap(
    db: AsyncSession,
    *,
    workspace_id: str,
    user_id: str,
    project_id: str | None,
    locale: str,
) -> AgentUiBootstrap:
    targets, default_scope = await execution_target_catalog(
        db,
        workspace_id=workspace_id,
        user_id=user_id,
        project_id=project_id,
    )
    chinese = locale.lower().startswith("zh")
    return AgentUiBootstrap(
        capabilities=default_ui_capabilities(),
        execution_targets=targets,
        execution_scope=default_scope,
        starter_prompts=_starter_prompts(project_id=project_id, chinese=chinese),
        composer_hint=(
            "输入 / 选择技能，或添加文件、工作流和运行记录作为上下文"
            if chinese
            else "Type / to choose a skill, or add files, workflows, and runs as context"
        ),
    )


def _starter_prompts(
    *, project_id: str | None, chinese: bool
) -> list[StarterPromptView]:
    if project_id and chinese:
        return [
            StarterPromptView(
                id="check-workflow",
                title="检查并运行工作流",
                prompt="检查当前项目的工作流和输入，确认无误后运行它。",
                icon="check",
            ),
            StarterPromptView(
                id="explain-inputs",
                title="解释项目输入",
                prompt="解释当前项目可用的工作流输入，以及应该如何准备数据。",
                icon="explain",
            ),
            StarterPromptView(
                id="review-run",
                title="复盘最近一次运行",
                prompt="查看当前项目最近一次运行，概括结果并指出需要处理的问题。",
                icon="review",
            ),
        ]
    if project_id:
        return [
            StarterPromptView(
                id="check-workflow",
                title="Check and run the workflow",
                prompt="Inspect this project's workflow and inputs, then run it when everything is ready.",
                icon="check",
            ),
            StarterPromptView(
                id="explain-inputs",
                title="Explain the project inputs",
                prompt="Explain the workflow inputs available in this project and how I should prepare the data.",
                icon="explain",
            ),
            StarterPromptView(
                id="review-run",
                title="Review the latest run",
                prompt="Review the latest run in this project, summarize the result, and identify anything that needs attention.",
                icon="review",
            ),
        ]
    if chinese:
        return [
            StarterPromptView(
                id="inspect-workspace",
                title="检查工作区",
                prompt="检查当前工作区，告诉我有哪些项目、工作流和最近的运行记录。",
                icon="check",
            ),
            StarterPromptView(
                id="plan-analysis",
                title="规划一次分析",
                prompt="帮我规划一次新的生物信息分析，并列出开始前需要准备的内容。",
                icon="explain",
            ),
            StarterPromptView(
                id="review-failures",
                title="检查失败运行",
                prompt="查找最近失败的运行，解释最可能的原因并给出下一步建议。",
                icon="review",
            ),
        ]
    return [
        StarterPromptView(
            id="inspect-workspace",
            title="Inspect the workspace",
            prompt="Inspect the current workspace and summarize its projects, workflows, and recent runs.",
            icon="check",
        ),
        StarterPromptView(
            id="plan-analysis",
            title="Plan an analysis",
            prompt="Help me plan a new bioinformatics analysis and list what I need before starting.",
            icon="explain",
        ),
        StarterPromptView(
            id="review-failures",
            title="Review failed runs",
            prompt="Find recent failed runs, explain the most likely causes, and recommend next steps.",
            icon="review",
        ),
    ]


__all__ = ["build_agent_ui_bootstrap"]

File: frontend/lib/agent/bootstrap.ts
import { apiRequest } from "@/lib/api"

export type AgentExecutionScope = {
  mode: "auto" | "manual"
  targetIds: string[]
}

export type AgentExecutionTarget = {
  id: string
  handle: string
  alias: string
  kind: "local" | "remote_ssh"
  status: "online" | "offline" | "error" | "unknown"
  primary: boolean
  disabledReason: string | null
}

export type AgentStarterPrompt = {
  id: string
  title: string
  prompt: string
  icon: "check" | "explain" | "review" | "chat"
}

export type AgentUiCapabilities = {
  reasoning: boolean
  toolActivity: boolean
  approvals: boolean
  artifacts: boolean
  starterPrompts: boolean
  multiTargetExecution: boolean
  retry: boolean
  editAndResend: boolean
}

export type AgentUiBootstrap = {
  protocolVersion: 1
  capabilities: AgentUiCapabilities
  executionTargets: AgentExecutionTarget[]
  executionScope: AgentExecutionScope
  starterPrompts: AgentStarterPrompt[]
  composerHint: string | null
  degradedReason: "unsupported_version" | "invalid_payload" | null
}

export async function getAgentUiBootstrap(
  projectId: string | null,
  locale: string,
): Promise<AgentUiBootstrap> {
  const response = await apiRequest<unknown>("/agent/ui/bootstrap", {
    params: { project_id: projectId ?? undefined, locale },
  })
  return normalizeAgentUiBootstrap(response.data, locale)
}

export function normalizeAgentUiBootstrap(
  value: unknown,
  locale: string,
): AgentUiBootstrap {
  if (!isRecord(value) || value.protocol_version !== 1) {
    return fallbackBootstrap(
      locale,
      isRecord(value) && typeof value.protocol_version === "number"
        ? "unsupported_version"
        : "invalid_payload",
    )
  }

  const targets = Array.isArray(value.execution_targets)
    ? value.execution_targets.flatMap(normalizeTarget)
    : []
  const scope = normalizeScope(value.execution_scope, targets)
  const prompts = Array.isArray(value.starter_prompts)
    ? value.starter_prompts.flatMap(normalizeStarterPrompt).slice(0, 4)
    : []
  const fallback = fallbackBootstrap(locale, null)

  return {
    protocolVersion: 1,
    capabilities: normalizeCapabilities(value.capabilities),
    executionTargets: targets.length > 0 ? targets : fallback.executionTargets,
    executionScope: scope ?? fallback.executionScope,
    starterPrompts: prompts.length > 0 ? prompts : fallback.starterPrompts,
    composerHint:
      typeof value.composer_hint === "string" && value.composer_hint.trim()
        ? value.composer_hint.trim()
        : fallback.composerHint,
    degradedReason: null,
  }
}

function normalizeCapabilities(value: unknown): AgentUiCapabilities {
  const record = isRecord(value) ? value : {}
  return {
    reasoning: record.reasoning === true,
    toolActivity: record.tool_activity === true,
    approvals: record.approvals === true,
    artifacts: record.artifacts === true,
    starterPrompts: record.starter_prompts === true,
    multiTargetExecution: record.multi_target_execution === true,
    retry: record.retry === true,
    editAndResend: record.edit_and_resend === true,
  }
}

function normalizeTarget(value: unknown): AgentExecutionTarget[] {
  if (!isRecord(value)) return []
  if (
    !isText(value.id, 200) ||
    !isText(value.handle, 200) ||
    !isText(value.alias, 200) ||
    (value.kind !== "local" && value.kind !== "remote_ssh")
  ) {
    return []
  }
  const status = ["online", "offline", "error", "unknown"].includes(
    String(value.status),
  )
    ? (value.status as AgentExecutionTarget["status"])
    : "unknown"
  return [
    {
      id: value.id,
      handle: value.handle,
      alias: value.alias,
      kind: value.kind,
      status,
      primary: value.primary === true,
      disabledReason:
        typeof value.disabled_reason === "string" ? value.disabled_reason : null,
    },
  ]
}

function normalizeScope(
  value: unknown,
  targets: AgentExecutionTarget[],
): AgentExecutionScope | null {
  if (!isRecord(value) || (value.mode !== "auto" && value.mode !== "manual")) {
    return null
  }
  const known = new Set(targets.map((target) => target.id))
  const targetIds = Array.isArray(value.target_ids)
    ? [...new Set(value.target_ids.filter((id): id is string => typeof id === "string" && known.has(id)))]
    : []
  if (value.mode === "manual" && targetIds.length === 0) return null
  return { mode: value.mode, targetIds }
}

function normalizeStarterPrompt(value: unknown): AgentStarterPrompt[] {
  if (!isRecord(value)) return []
  if (
    !isText(value.id, 100) ||
    !isText(value.title, 120) ||
    !isText(value.prompt, 2000)
  ) {
    return []
  }
  const icon = ["check", "explain", "review", "chat"].includes(String(value.icon))
    ? (value.icon as AgentStarterPrompt["icon"])
    : "chat"
  return [{ id: value.id, title: value.title, prompt: value.prompt, icon }]
}

function fallbackBootstrap(
  locale: string,
  degradedReason: AgentUiBootstrap["degradedReason"],
): AgentUiBootstrap {
  const chinese = locale.toLowerCase().startsWith("zh")
  return {
    protocolVersion: 1,
    capabilities: {
      reasoning: true,
      toolActivity: true,
      approvals: true,
      artifacts: true,
      starterPrompts: true,
      multiTargetExecution: false,
      retry: true,
      editAndResend: true,
    },
    executionTargets: [
      {
        id: "local",
        handle: "local",
        alias: chinese ? "本地" : "Local",
        kind: "local",
        status: "online",
        primary: true,
        disabledReason: null,
      },
    ],
    executionScope: { mode: "auto", targetIds: [] },
    starterPrompts: chinese
      ? [
          { id: "inspect-workspace", title: "检查工作区", prompt: "检查当前工作区并概括项目、工作流和最近运行。", icon: "check" },
          { id: "plan-analysis", title: "规划一次分析", prompt: "帮我规划一次新的生物信息分析，并列出需要准备的内容。", icon: "explain" },
          { id: "review-failures", title: "检查失败运行", prompt: "检查最近失败的运行，解释原因并给出下一步建议。", icon: "review" },
        ]
      : [
          { id: "inspect-workspace", title: "Inspect the workspace", prompt: "Inspect this workspace and summarize its projects, workflows, and recent runs.", icon: "check" },
          { id: "plan-analysis", title: "Plan an analysis", prompt: "Help me plan a bioinformatics analysis and list what I need before starting.", icon: "explain" },
          { id: "review-failures", title: "Review failed runs", prompt: "Review recent failed runs, explain likely causes, and recommend next steps.", icon: "review" },
        ],
    composerHint: chinese
      ? "输入 / 选择技能，或添加文件、工作流和运行记录作为上下文"
      : "Type / to choose a skill, or add files, workflows, and runs as context",
    degradedReason,
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value)
}

function isText(value: unknown, maxLength: number): value is string {
  return typeof value === "string" && value.trim().length > 0 && value.length <= maxLength
}
